horovod/common/operations.cc

```diff
@@ -818,135 +864,143 @@ void PerformOperation(TensorTable& tensor_table, MPIResponse response) {
         ACTIVITY_END_ALL(entries, timeline)
       }
 
-      ACTIVITY_START_ALL(entries, timeline, "SCHEDULE")
-
-      cudaEvent_t queue_end_event = nullptr;
       if (timeline.Initialized()) {
-        RECORD_EVENT(entries, queue_end_event, stream);
+        RECORD_EVENT(entries, event_queue, QUEUE, stream)
       }
 
-      cudaEvent_t after_memcpy_in_event = nullptr;
-      cudaEvent_t after_reduce_event = nullptr;
-      cudaEvent_t after_memcpy_out_event = nullptr;
+      // If entries.size() > 1, we copy tensors into fusion buffer before
+      // allreduce, and distribute results of allreduce back into target
+      // tensors after allreduce.
+
+      void* buffer_data;
+      int64_t num_elements = 0;
+      size_t buffer_len;
       if (entries.size() > 1) {
         // Access the fusion buffer.
         auto& buffer = horovod_global.tensor_fusion_buffers[std::make_tuple(
             first_entry.device, first_entry.context->framework())];
-        auto buffer_data = buffer->AccessData(first_entry.context);
+        buffer_data =
+            const_cast<void*>(buffer->AccessData(first_entry.context));
 
         // Copy memory into the fusion buffer.
         int64_t offset = 0;
-        for (auto it = entries.begin(); it != entries.end(); it++) {
+        for (auto& e : entries) {
           void* buffer_data_at_offset = (uint8_t*)buffer_data + offset;
           CUDA_CHECK(entries, "cudaMemcpyAsync",
-                     cudaMemcpyAsync(buffer_data_at_offset, it->tensor->data(),
-                                     (size_t)it->tensor->size(),
+                     cudaMemcpyAsync(buffer_data_at_offset, e.tensor->data(),
+                                     (size_t)e.tensor->size(),
                                      cudaMemcpyDeviceToDevice, stream))
-          offset += it->tensor->size();
+          offset += e.tensor->size();
         }
+        buffer_len = (size_t)offset;
         if (timeline.Initialized()) {
-          RECORD_EVENT(entries, after_memcpy_in_event, stream)
+          RECORD_EVENT(entries, event_queue, MEMCPY_IN_FUSION_BUFFER, stream)
         }
 
         // Perform the reduction on the fusion buffer.
-        int64_t num_elements = 0;
-        for (auto it = entries.begin(); it != entries.end(); it++) {
-          num_elements += it->tensor->shape().num_elements();
+        for (auto& e : entries) {
+          num_elements += e.tensor->shape().num_elements();
         }
-        NCCL_CHECK(entries, "ncclAllReduce",
-                   ncclAllReduce(buffer_data, (void*)buffer_data,
-                                 (size_t)num_elements,
-                                 GetNCCLDataType(first_entry.tensor), ncclSum,
-                                 nccl_comm, stream))
+      } else {
+        buffer_data = (void*)first_entry.output->data();
```

Comment:
@alsrgv It looks like a bug.
If (entries.size() == 1), you set output buffer "buffer_data" to "first_entry.output->data()" and then you call "ncclReduce(buffer_data, buffer_data, ...)". You need to copy input data "first_entry.tensor->data()" to "buffer_data" before the ncclReduce. Or, you need a different ncclReduce call taking the input buffer for this special case.

Reply:
@nvcastet, great catch, thanks a lot!  Pushing through a PR #358 with the fix.